File: lib/pricing-engine/scenario.ts
import { NSSParameters } from "./types";
import { ZeroCurve, bootstrapZeroCurve } from "./bootstrap";

export function nssYield(
  t: number,
  beta0: number,
  beta1: number,
  beta2: number,
  beta3: number,
  tau1: number,
  tau2: number
): number {
  if (t === 0) return beta0 + beta1;
  const term1 = (1.0 - Math.exp(-t / tau1)) / (t / tau1);
  const term2 = term1 - Math.exp(-t / tau1);
  const term3 = (1.0 - Math.exp(-t / tau2)) / (t / tau2) - Math.exp(-t / tau2);
  return beta0 + beta1 * term1 + beta2 * term2 + beta3 * term3;
}

export function applyScenarioShocks(
  baseParams: NSSParameters,
  shocks: {
    parallel_shift?: number;
    slope_shock?: number;
    curvature1_shock?: number;
    curvature2_shock?: number;
    twist_shock?: number;
    twist_pivot?: number;
  }
): NSSParameters {
  const parallelShift = shocks.parallel_shift ?? 0.0;
  const slopeShock = shocks.slope_shock ?? 0.0;
  const curvature1Shock = shocks.curvature1_shock ?? 0.0;
  const curvature2Shock = shocks.curvature2_shock ?? 0.0;
  const twistShock = shocks.twist_shock ?? 0.0;
  const twistPivot = shocks.twist_pivot ?? 5.0;

  const { beta0, beta1, beta2, beta3, tau1, tau2 } = baseParams;

  // Calculate twist offset at pivot
  let g1Pivot = 0.0;
  if (twistPivot > 0) {
    g1Pivot = (1.0 - Math.exp(-twistPivot / tau1)) / (twistPivot / tau1);
  } else {
    g1Pivot = 1.0;
  }
  const deltaBeta0Twist = -twistShock * g1Pivot;

  let newBeta0 = beta0 + parallelShift + deltaBeta0Twist;
  let newBeta1 = beta1 + slopeShock + twistShock;
  let newBeta2 = beta2 + curvature1Shock;
  let newBeta3 = beta3 + curvature2Shock;

  // Clamp level parameter beta0
  newBeta0 = Math.max(0.0, Math.min(25.0, newBeta0));

  return {
    beta0: newBeta0,
    beta1: Math.max(-25.0, Math.min(25.0, newBeta1)),
    beta2: Math.max(-25.0, Math.min(25.0, newBeta2)),
    beta3: Math.max(-25.0, Math.min(25.0, newBeta3)),
    tau1,
    tau2
  };
}

export function getShockedZeroCurve(
  baseParams: NSSParameters,
  shocks: {
    parallel_shift?: number;
    slope_shock?: number;
    curvature1_shock?: number;
    curvature2_shock?: number;
    twist_shock?: number;
    twist_pivot?: number;
  },
  maxMaturity: number = 40.0,
  stepSize: number = 0.5
): ZeroCurve {
  const shockedParams = applyScenarioShocks(baseParams, shocks);
  
  const parCurveFn = (t: number) => {
    return nssYield(
      t,
      shockedParams.beta0,
      shockedParams.beta1,
      shockedParams.beta2,
      shockedParams.beta3,
      shockedParams.tau1,
      shockedParams.tau2
    );
  };
  
  return bootstrapZeroCurve(parCurveFn, maxMaturity, stepSize);
}

File: backend/quant_core/scenario.py
import numpy as np
from quant_core.nss import nss_yield
from quant_core.bootstrap import bootstrap_zero_curve, build_zero_curve_from_zero_rates

def apply_scenario_shocks(
    base_params: dict,
    parallel_shift: float = 0.0,
    slope_shock: float = 0.0,
    curvature1_shock: float = 0.0,
    curvature2_shock: float = 0.0,
    twist_shock: float = 0.0,
    twist_pivot: float = 5.0
) -> dict:
    """
    Applies factor shocks to base Nelson-Siegel-Svensson parameters.
    """
    beta0 = base_params["beta0"]
    beta1 = base_params["beta1"]
    beta2 = base_params["beta2"]
    beta3 = base_params["beta3"]
    tau1 = base_params["tau1"]
    tau2 = base_params["tau2"]
    
    # Calculate twist offset to preserve rate at twist_pivot
    g1_pivot = 0.0
    if twist_pivot > 0:
        g1_pivot = (1.0 - np.exp(-twist_pivot / tau1)) / (twist_pivot / tau1)
    else:
        g1_pivot = 1.0
        
    delta_beta0_twist = -twist_shock * g1_pivot
    
    new_beta0 = beta0 + parallel_shift + delta_beta0_twist
    new_beta1 = beta1 + slope_shock + twist_shock
    new_beta2 = beta2 + curvature1_shock
    new_beta3 = beta3 + curvature2_shock
    
    # Enforce boundaries on beta0 (yield level cannot become negative or implausibly high)
    new_beta0 = max(0.0, min(25.0, new_beta0))
    
    return {
        "beta0": new_beta0,
        "beta1": max(-25.0, min(25.0, new_beta1)),
        "beta2": max(-25.0, min(25.0, new_beta2)),
        "beta3": max(-25.0, min(25.0, new_beta3)),
        "tau1": tau1,
        "tau2": tau2
    }

def get_shocked_zero_curve(
    base_params: dict,
    parallel_shift: float = 0.0,
    slope_shock: float = 0.0,
    curvature1_shock: float = 0.0,
    curvature2_shock: float = 0.0,
    twist_shock: float = 0.0,
    twist_pivot: float = 5.0,
    max_maturity: float = 40.0,
    step_size: float = 0.5,
    yield_type: str = "par"
):
    """
    Returns a new ZeroCurve after applying the NSS factor shocks and bootstrapping.
    """
    shocked_params = apply_scenario_shocks(
        base_params=base_params,
        parallel_shift=parallel_shift,
        slope_shock=slope_shock,
        curvature1_shock=curvature1_shock,
        curvature2_shock=curvature2_shock,
        twist_shock=twist_shock,
        twist_pivot=twist_pivot
    )
    
    def zc_fn(t):
        return nss_yield(
            t,
            shocked_params["beta0"],
            shocked_params["beta1"],
            shocked_params["beta2"],
            shocked_params["beta3"],
            shocked_params["tau1"],
            shocked_params["tau2"]
        )
        
    if yield_type == "par":
        return bootstrap_zero_curve(zc_fn, max_maturity=max_maturity, step_size=step_size)
    return build_zero_curve_from_zero_rates(zc_fn, max_maturity=max_maturity, step_size=step_size)
